cli/cmd/routes.go

```diff
@@ -0,0 +1,281 @@
+package cmd
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"sort"
+	"strings"
+	"text/tabwriter"
+	"time"
+
+	"github.com/linkerd/linkerd2/controller/api/util"
+	pb "github.com/linkerd/linkerd2/controller/gen/public"
+	"github.com/linkerd/linkerd2/pkg/k8s"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+)
+
+type routesOptions struct {
+	namespace     string
+	timeWindow    string
+	fromNamespace string
+	fromResource  string
+	outputFormat  string
+}
+
+func newRoutesOptions() *routesOptions {
+	return &routesOptions{
+		namespace:     "default",
+		timeWindow:    "1m",
+		fromNamespace: "",
+		fromResource:  "",
+		outputFormat:  "",
+	}
+}
+
+func newCmdRoutes() *cobra.Command {
+	options := newRoutesOptions()
+
+	cmd := &cobra.Command{
+		Use:   "routes [flags] (SERVICE)",
+		Short: "Display route stats about a service",
+		Long: `Display route stats about a service.
+
+This command will only work for services that have a Service Profile defined.`,
+		Example: `  # Routes for the webapp service in the test namespace.
+  linkerd routes webapp -n test
+
+  # Routes for calls from from the traffic deployment to the webapp service in the test namespace.
+  linkerd routes webapp -n test --from deploy/traffic --from-namespace test`,
+		Args:      cobra.ExactArgs(1),
+		ValidArgs: util.ValidTargets,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			req, err := buildTopRoutesRequest(args[0], options)
+			if err != nil {
+				return fmt.Errorf("error creating metrics request while making routes request: %v", err)
+			}
+
+			output, err := requestRouteStatsFromAPI(validatedPublicAPIClient(time.Time{}), req, options)
+			if err != nil {
+				return err
+			}
+
+			_, err = fmt.Print(output)
+
+			return err
+		},
+	}
+
+	cmd.PersistentFlags().StringVarP(&options.namespace, "namespace", "n", options.namespace, "Namespace of the specified resource")
+	cmd.PersistentFlags().StringVarP(&options.timeWindow, "time-window", "t", options.timeWindow, "Stat window (for example: \"10s\", \"1m\", \"10m\", \"1h\")")
+	cmd.PersistentFlags().StringVar(&options.fromResource, "from", options.fromResource, "If present, restricts outbound stats from the specified resource name")
```

Comment:
Note that the `--to` flag is not supported here because the target of a routes command must be a service.